### PR TITLE
ENH: Add IndicatorBox functional

### PR DESCRIPTION
**Pull request by adler-j**
Cherry picked from #587, should be simple to review.

Further changes to be done:

* Add `IndicatorNonnegativity`
* Add test

### CODE REVIEW

**Comment by adler-j** [comment]
Needs a test

File: odl/solvers/functional/default_functionals.py [hunk 1]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
+        >>> func([0, 1, 3])  # one point outside
+        inf
+        """
+        Functional.__init__(self, space, linear=False)
+        self.lower = lower
+        self.upper = upper
+
+    def _call(self, x):
+        """Apply the functional to the given point."""
+        out = self.domain.element()
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Explain what magic happens here, also perhaps optimize for the case where the bounds are scalar.

File: odl/solvers/functional/default_functionals.py [hunk 2]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
+            x.ufunc.minimum(self.upper, out=out)
+        elif self.lower is not None and self.upper is not None:
+            x.ufunc.maximum(self.lower, out=out)
+            out.ufunc.minimum(self.upper, out=out)
+        else:
+            out.assign(x)
+
+        dist = x.dist(out)
+
+        if dist > 0:
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Just do `if x.dist(out) > 0` here

Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Or even

```
return np.inf if x.dist(out) > 0 else 0
```

File: odl/solvers/functional/default_functionals.py [hunk 3]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = IndicatorBox(space, 0, 2)
+        >>> func([0, 1, 2])  # all points inside
+        0
+        >>> func([0, 1, 3])  # one point outside
+        inf
+        """
+        Functional.__init__(self, space, linear=False)
+        self.lower = lower
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: `self.lower = space.element(lower)` + special case for scalar. See the proximal for how to

File: odl/solvers/functional/default_functionals.py [hunk 4]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
+            Domain of the functional G(x)
+        lower : ``space.field`` element or ``space`` `element-like`, optional
+            The lower bound.
+            Default: ``None``, interpreted as -infinity
+        upper : ``space.field`` element or ``space`` `element-like`, optional
+            The upper bound.
+            Default: ``None``, interpreted as +infinity
+
+        Examples
+        --------
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Short string on what happens

File: odl/solvers/functional/default_functionals.py [hunk 5]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
+
+    """Indicator on some box shaped domain."""
+
+    def __init__(self, space, lower, upper):
+        """Initialize an instance
+
+        Parameters
+        ----------
+        space : `LinearSpace`
+            Domain of the functional G(x)
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: What is G?

File: odl/solvers/functional/default_functionals.py [hunk 6]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
...
         """
         super().__init__(space=space, constant=0)
+
+
+class IndicatorBox(Functional):
+
+    """Indicator on some box shaped domain."""
+
+    def __init__(self, space, lower, upper):
+        """Initialize an instance
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Missing dot

File: odl/solvers/functional/default_functionals.py [hunk 7]
```diff
@@ -464,3 +464,67 @@ def __init__(self, space):
             Domain of the functional.
         """
         super().__init__(space=space, constant=0)
+
+
+class IndicatorBox(Functional):
+
+    """Indicator on some box shaped domain."""
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Should be clear in what sense this is a box (i.e. in the VALUES) of the incoming function.

**Comment by adler-j** [comment]
Further changes to be done:

* Add `IndicatorNonnegativity`

**Comment by kohr-h** [comment]
See #587 for review comments.

**Comment by kohr-h** [comment]
@adler-j  Please write a release note on this.

**Pull request merged**